src/tool_cb_prg.c

```diff
@@ -117,34 +118,10 @@ void progressbarinit(struct ProgressData *bar,
   if(config->use_resume)
     bar->initial_size = config->resume_from;
 
-/* TODO: get terminal width through ansi escapes or something similar.
-   try to update width when xterm is resized... - 19990617 larsa */
-#ifndef __EMX__
-  /* 20000318 mgs
-   * OS/2 users most likely won't have this env var set, and besides that
-   * we're using our own way to determine screen width */
-  colp = curlx_getenv("COLUMNS");
-  if(colp) {
-    char *endptr;
-    long num = strtol(colp, &endptr, 10);
-    if((endptr != colp) && (endptr == colp + strlen(colp)) && (num > 0))
-      bar->width = (int)num;
-    else
-      bar->width = 79;
-    curl_free(colp);
-  }
-  else
-    bar->width = 79;
```

Comment:
I would leave all the existing code and insert with guard

~~~c
...
#if defined(HAVE_SYS_IOCTL_H) || defined(HAVE_UNISTD_H)
else if(!ioctl(STDOUT_FILENO, TIOCGWINSZ, &w))
  bar->width = w.ws_col - 1;
#endif
else
  bar->width = 79;
~~~

and at the top you'll have to guard as well. the reason is sys/ioctl unistd is not guaranteed.
~~~c
#ifdef HAVE_SYS_IOCTL_H
#  include <sys/ioctl.h>
#endif
#ifdef HAVE_UNISTD_H
#  include <unistd.h>
#endif
~~~

also you might need termio
~~~c
#ifdef HAVE_TERMIOS_H
#  include <termios.h>
#elif defined(HAVE_TERMIO_H)
#  include <termio.h>
#endif
~~~

Reply:
Thanks, I updated the headers, but why would I need `termio`?

Reply:
I keep having errors about `ioctl` not being defined if I had the `#ifdef`.

```c
tool_cb_prg.c:113:18: error: variable has incomplete type 'struct winsize'
  struct winsize w;
                 ^
tool_cb_prg.c:113:10: note: forward declaration of 'struct winsize'
  struct winsize w;
         ^
tool_cb_prg.c:127:12: warning: implicit declaration of function 'ioctl' is invalid in
      C99 [-Wimplicit-function-declaration]
  else if(!ioctl(STDOUT_FILENO, TIOCGWINSZ, &w)) {
           ^
tool_cb_prg.c:127:33: error: use of undeclared identifier 'TIOCGWINSZ'
  else if(!ioctl(STDOUT_FILENO, TIOCGWINSZ, &w)) {
                                ^
tool_cb_prg.c:128:26: error: use of undeclared identifier 'TIOCGWINSZ'
    ioctl(STDOUT_FILENO, TIOCGWINSZ, &w);
```

Even with the `ifdef`:

```c
#if defined(HAVE_SYS_IOCTL_H) && defined(HAVE_UNISTD_H)
  else if(!ioctl(STDOUT_FILENO, TIOCGWINSZ, &w)) {
    ioctl(STDOUT_FILENO, TIOCGWINSZ, &w);
    return w.ws_col - 1;
  }
#endif
```

Reply:
termio for TIOCGWINSZ. There's still no reason to modify so many lines. The function was fine as is,I see no reason why you couldn't just add your code in several lines. It looks like you changed it to repeatedly get the window size in case someone resizes the terminal while downloading. I think that's kind of unnecessary, and may add some unwanted overhead